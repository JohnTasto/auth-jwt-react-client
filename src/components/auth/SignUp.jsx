import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Field, reduxForm } from 'redux-form'
import { v4 } from 'uuid'
import * as actions from '../../ducks/auth'
import {
  requiredValidator,
  emailValidator,
  passwordValidator,
} from '../../helpers/validators'


const renderField = ({ input, label, type, meta: { touched, error } }) => {
  const id = v4()
  return (
    <fieldset className="form-group">
      <label htmlFor={id}>{label}</label>
      <div>
        <input id={id} {...input} placeholder={label} type={type} className="form-control" />
        {touched && error && <span className="error">{error}</span>}
      </div>
    </fieldset>
  )
}

export class SignUp extends Component {
  componentWillMount() {
    const { location, redirectLocation, setAuthRedirect } = this.props
    if (!redirectLocation) {
      if (location.state && location.state.from) {
        setAuthRedirect(location.state.from)
      } else {
        setAuthRedirect({ pathname: '/' })
      }
    }
  }

  renderAlert() {
    if (this.props.error) {
      return (
        <div className="alert alert-danger">
          <strong>Error:</strong> {this.props.error}
        </div>
      )
    }
    return ''
  }

  render() {
    const { authenticated, handleSubmit, invalid, submitting, submitSucceeded } = this.props

    if (authenticated) {
      return (
        <div>
          <p>You are already signed in.</p>
          <button onClick={this.props.signOut}>Sign Out</button>
        </div>
      )
    }

    if (submitSucceeded) {
      return (
        <div>
          <p>Thank you for signing up. Check your email to verify your account.</p>
        </div>
      )
    }

    return (
      <form onSubmit={handleSubmit(this.props.signUp)}>
        <Field
          name="email"
          type="email"
          component={renderField}
          label="Email"
          validate={[requiredValidator, emailValidator]}
        />
        <Field
          name="password"
          type="password"
          component={renderField}
          label="Password"
          validate={[requiredValidator, passwordValidator]}
        />
        <Field
          name="passwordConfirm"
          type="password"
          component={renderField}
          label="Confirm Password"
          validate={requiredValidator}
        />
        {this.renderAlert()}
        <button action="submit" className="btn btn-primary" disabled={invalid || submitting}>
          Sign up!
        </button>
      </form>
    )
  }
}

function validate(values) {
  const errors = {}

  if (values.password !== values.passwordConfirm) {
    errors.passwordConfirm = 'Passwords must match'
  }

  return errors
}

export default connect(
  state => ({
    authenticated: state.auth.authenticated,
    errorMessage: state.auth.error,
    redirectLocation: state.auth.redirectLocation,
  }),
  actions,
  (...props) =>
    Object.assign({}, ...props),
)(reduxForm({
  form: 'signUp',
  validate,
})(SignUp))
