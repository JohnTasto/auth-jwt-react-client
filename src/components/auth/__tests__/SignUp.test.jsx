/* eslint-env jest */
/* eslint-disable comma-dangle */

import React from 'react'
import { StaticRouter, Route } from 'react-router-dom'
import { Provider } from 'react-redux'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import createStore from '../../../store'
import SignUp, { SignUp as SignUpComp } from '../SignUp'


describe('<SignUp />', () => {

  test('Renders form if not authenticated', () => {
    const wrapper = shallow(
      <SignUpComp
        handleSubmit={jest.fn()}
        setAuthRedirect={jest.fn()}
        location={{}}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Renders sign out button if authenticated', () => {
    const wrapper = shallow(
      <SignUpComp
        authenticated
        handleSubmit={jest.fn()}
        setAuthRedirect={jest.fn()}
        location={{}}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Renders success message if submit succeeded', () => {
    const wrapper = shallow(
      <SignUpComp
        submitSucceeded
        handleSubmit={jest.fn()}
        setAuthRedirect={jest.fn()}
        location={{}}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  describe('Integration', () => {
    const emailError = 'input[name="email"] ~ .error'
    const passwordError = 'input[name="password"] ~ .error'
    const passwordConfirmError = 'input[name="passwordConfirm"] ~ .error'

    let wrapper
    let props
    let submitButton

    beforeEach(() => {
      props = {
        signUp: jest.fn(),
      }
      wrapper = mount(
        <Provider store={createStore()}>
          <StaticRouter location="/signup" context={{}} >
            <Route
              path="/signup"
              render={routeProps => (<SignUp {...routeProps} {...props} />)}
            />
          </StaticRouter>
        </Provider>
      )
      submitButton = wrapper.find('[action="submit"]')
    })


    test('Calls action signUp with specified email and password', () => {
      const user = {
        email: 'e@m.ail',
        password: 'Password1',
        passwordConfirm: 'Password1',
      }

      // should have no errors, submit button disabled
      expect(wrapper.find(emailError).length).toBe(0)
      expect(wrapper.find(passwordError).length).toBe(0)
      expect(wrapper.find(passwordConfirmError).length).toBe(0)
      expect(submitButton.props().disabled).toBe(true)

      // simulate valid form fill
      wrapper.find('[name="email"]').simulate('change', { target: { value: user.email } })
      wrapper.find('[name="password"]').simulate('change', { target: { value: user.password } })
      wrapper.find('[name="passwordConfirm"]').simulate('change', { target: { value: user.passwordConfirm } })

      // should still have no errors, submit button enabled
      expect(wrapper.find(emailError).length).toBe(0)
      expect(wrapper.find(passwordError).length).toBe(0)
      expect(wrapper.find(passwordConfirmError).length).toBe(0)
      expect(submitButton.props().disabled).toBe(false)

      // simulate form submit
      wrapper.find('form').simulate('submit')

      // verify result
      expect(props.signUp.mock.calls[0][0]).toEqual(user)
    })


    test('Displays error and disables submit button on empty input', () => {

      // should have no errors
      expect(wrapper.find(emailError).length).toBe(0)
      expect(wrapper.find(passwordError).length).toBe(0)
      expect(wrapper.find(passwordConfirmError).length).toBe(0)
      expect(submitButton.props().disabled).toBe(true)

      // simulate empty form fill
      wrapper.find('input[name="email"]').simulate('blur')
      wrapper.find('input[name="password"]').simulate('blur')
      wrapper.find('input[name="passwordConfirm"]').simulate('blur')

      // should now have errors
      expect(wrapper.find(emailError).length).toBe(1)
      expect(wrapper.find(passwordError).length).toBe(1)
      expect(wrapper.find(passwordConfirmError).length).toBe(1)
      expect(submitButton.props().disabled).toBe(true)
    })


    test('Displays error on malformed email', () => {

      // should have no error
      expect(wrapper.find(emailError).length).toBe(0)
      expect(submitButton.props().disabled).toBe(true)

      // simulate entering malformed email
      wrapper.find('input[name="email"]')
        .simulate('change', { target: { value: 'email' } })
        .simulate('blur')

      // should now have error
      expect(wrapper.find(emailError).length).toBe(1)
      expect(submitButton.props().disabled).toBe(true)
    })


    test('Displays error on malformed password', () => {

      // should have no error
      expect(wrapper.find(passwordError).length).toBe(0)
      expect(submitButton.props().disabled).toBe(true)

      // simulate entering malformed password
      wrapper.find('input[name="password"]')
        .simulate('change', { target: { value: 'password' } })
        .simulate('blur')

      // should now have error
      expect(wrapper.find(passwordError).length).toBe(1)
      expect(submitButton.props().disabled).toBe(true)
    })


    test('Displays error if passwords do not match', () => {

      // should have no error
      expect(wrapper.find(passwordConfirmError).length).toBe(0)
      expect(submitButton.props().disabled).toBe(true)

      // simulate entering valid, but non-matching passwords
      wrapper.find('input[name="password"]')
        .simulate('change', { target: { value: 'Password1' } })
        .simulate('blur')
      wrapper.find('input[name="passwordConfirm"]')
        .simulate('change', { target: { value: 'Password2' } })
        .simulate('blur')

      // should now have error
      expect(wrapper.find(passwordConfirmError).length).toBe(1)
      expect(submitButton.props().disabled).toBe(true)
    })
  })
})
