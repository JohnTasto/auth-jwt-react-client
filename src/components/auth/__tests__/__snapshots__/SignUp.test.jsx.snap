// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> Renders form if not authenticated 1`] = `
<form>
  <Field
    component={[Function]}
    label="Email"
    name="email"
    type="email"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="Password"
    name="password"
    type="password"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="Confirm Password"
    name="passwordConfirm"
    type="password"
    validate={[Function]}
  />
  <button
    action="submit"
    className="btn btn-primary"
  >
    Sign up!
  </button>
</form>
`;

exports[`<SignUp /> Renders sign out button if authenticated 1`] = `
<div>
  <p>
    You are already signed in.
  </p>
  <button>
    Sign Out
  </button>
</div>
`;
